Command Tool `[AARD-2079]`

AlexD717 opened the pull request:
## Task
Create a command tool, that allows synthesis users to type in "/" followed by a command that then executes a specific action. 
<!--
Please include any relevant Jira ticket ID(s) at the end of the PR title, in the form AARD-xxxx, where "AARD" is Jira project.
Include the same Jira ticket ID(s) in this section.
-->

AARD-2079

<!--
Provide a brief description of what the task was here.
-->

## Symptom
It can get annoying for developers and users to have to go through multiple panels to achive what they want. For advanced users of Synthesis a command tool would let them quickly type commands to implement a specific action. For example, instead of having to go to debug tools, then clicking toggle drag mode, then clicking hide debug tools panel, you could just type "/drag mode" and it will toggle the drag mode.
<!--
How does the problem manifest itself?
Describe the problem as seen by the user, or by the caller or the code if it's not directly visible to the user.

Note: "Symptom" can include new product use cases, not just bugs.
-->

## Solution
Implement a simple Command palette that appears when pressing "/" and lists options of different commands. It uses an algorithm to determine what you are most likely typing and then suggesting that as an option. This means that both "/drag" or "/drqg" (typo example) will triggle the toggle drag mode command.
<!--
How did you fix the problem/symptom?
Explain your approach and reasoning for choosing this solution.
-->

## Verification

- Pressing "/" opens up a command pallet
  - Doesn't open in the welcome screen (singleplayer/multiplayer selection screen)
  - Doesn't open if a modal is open
- Pressing "esc" closes the command palette
- Starting to type in a command has different options appear, with options that you most likely want appearing at the bottom
- pressing enter runs the command you have selected (or clicking on the command with the mouse)
- Can use arrow keys to navigate between selected commands (or mouse)
- Clicking outside of the command palette closes it

<!--
How did you test and verify your changes were correct?
List steps taken, tests/added/updated, and any manual verification done that should be replicated in review.
-->

---

Before merging, ensure the following criteria are met:

- [ ] All acceptance criteria outlined in the ticket are met.
- [ ] Necessary test cases have been added and updated.
- [ ] A feature toggle or safe disable path has been added (if applicable).
- [ ] User-facing polish:
  - Ask: *"Is this ready-looking?"*
- [ ] Cross-linking between Jira and GitHub:
  - PR links to the relevant Jira issue.
  - Jira ticket has a comment referencing this PR.

AlexD717 committed:
Simple Command Tool

AlexD717 committed:
Toggles Drag Mode & Cleaner Code

AlexD717 committed:
Smarter Search

AlexD717 committed:
General UX Improvements

azaleacolburn commented:
> Doesn't open in menu selection

1. I'm not sure what you mean by this, but it does still open when you have panels open (including the main menu panel which is what I assume you mean). If you mean within the command palette, then yes indeed typing `/` doesn't open another one.
2. I like how the drag mode command works, I think the commands that open panels should toggle them instead.
3. I think that trying to navigate past the first or last option should wrap around to the other end of the list.

AlexD717 commented:
> > Doesn't open in menu selection
> 
> 1. I'm not sure what you mean by this, but it does still open when you have panels open (including the main menu panel which is what I assume you mean). If you mean within the command palette, then yes indeed typing `/` doesn't open another one.

I meant the welcome screen. Will clarify that
<img width="276" height="332" alt="Screenshot 2025-08-14 at 10 38 03 AM" src="https://github.com/user-attachments/assets/f1105c22-0de0-4f65-9abd-73d007eb9161" />

I agree with your other suggestions and will implement them

AlexD717 committed:
Arrow Keys Cycle

AlexD717 commented:
> 2. I like how the drag mode command works, I think the commands that open panels should toggle them in

Waiting on pr #1268 to be merged in first, as that implements support for detecting if panels/modals are already opened

AlexD717 committed:
Commands for Configuring Robots/Fields

AlexD717 committed:
More Commands

rutmanz committed:
fix: remove casting

rutmanz commented on fission/src/ui/components/CommandPalette.tsx:
```diff
@@ -0,0 +1,374 @@
+
+    const openImportPanel = useCallback(
+        (configurationType: ConfigurationType) => {
+            openPanel<void, { configurationType: ConfigurationType }>(ImportMirabufPanel, { configurationType })
```
```suggestion
            openPanel(ImportMirabufPanel, { configurationType })
```

rutmanz reviewed (changes requested):
1. Looks like typing in the search bar also drives robots around (WASD and arrow keys)
2. I think whenever you reference a robot name, you should also reference the control scheme name (currently two dozers can't be distinguished)

AlexD717 committed:
Update fission/src/ui/components/CommandPalette.tsx

Co-authored-by: Zach Rutman <zarutman+git@gmail.com>

AlexD717 committed:
Include Robot Control Scheme Name

AlexD717 committed:
Disables Input if Command Palette is Open

rutmanz reviewed (changes requested):
The Luna scheme conflicts with this binding. I would either make the triggering key rebindable or prevent inputs from using it

AlexD717 committed:
Switch Luna Default Inputs

BrandonPacewic reviewed (changes requested):
I’d prefer if we set up a system where other parts of the simulator can register their own commands rather than hard-coding a single inline list here. That way each feature owns its own actions, and the palette just queries a shared registry.

I’m fine with keeping the inline list temporarily so we can get this merged without blocking other work, but we should make an effort to lay out how this will be adopted within this current PR.

AlexD717 committed:
Separate CommandRegistry

AlexD717 committed:
Commands Registered by Proper Objects

AlexD717 committed:
Formatting

AlexD717 committed:
Lint Error Fix

AlexD717 commented:
> I’d prefer if we set up a system where other parts of the simulator can register their own commands rather than hard-coding a single inline list here. That way each feature owns its own actions, and the palette just queries a shared registry.
> 

Added support for this, and moved the commands that made sense to the appropriate files

itspvty reviewed (approved):
I really like this concept overall and the use of fuse.js is very cool.

ryanzhangofficial reviewed (approved):
love it!

AlexD717 committed:
Merge branch 'dev' into alexey/2079/command-tool

AlexD717 committed:
Command Pallete Analytics

BrandonPacewic committed:
Merge branch 'dev' into alexey/2079/command-tool

PepperLola reviewed (changes requested):
Could you get rid of this highlight ring?

<img width="820" height="134" alt="Image" src="https://github.com/user-attachments/assets/846e2e07-f038-4e9c-b0a6-23b399721eb6" />

Also, I think it would be worth adding options for the different tabs of the configure panel just like there are for spawning.

AlexD717 commented:
> Could you get rid of this highlight ring?
> 
> <img alt="Image" width="820" height="134" src="https://private-user-images.githubusercontent.com/20529380/480816403-846e2e07-f038-4e9c-b0a6-23b399721eb6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU4NzY2NDgsIm5iZiI6MTc1NTg3NjM0OCwicGF0aCI6Ii8yMDUyOTM4MC80ODA4MTY0MDMtODQ2ZTJlMDctZjAzOC00ZTljLWIwYTYtMjNiMzk5NzIxZWI2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODIyVDE1MjU0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlkMjBkNjc1ZmEzNGQ1MWU5NjkxNjlkM2IwMDExYWMzZmU2NzJhODk2M2ZiYjdlYzZjMWVmZTI2NmFmNzY5OTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.t5Ww6kQtwdvVXP7Y8J6fUkcc8grN_Kar9tsBvLDJ6zE">

In my opinion, the command pallet without any border looks weird, when you have something other than the bottom part selected.
<img width="605" height="277" alt="Screenshot 2025-08-22 082846" src="https://github.com/user-attachments/assets/8200354b-1c54-4140-9bae-6aaef1bb1d5e" />


If I make the boarder 1px dark grey, it adds a slight line, which separates the text box, making it more clear. 
<img width="607" height="277" alt="Screenshot 2025-08-22 083021" src="https://github.com/user-attachments/assets/9ec3c078-9c13-493f-8786-0d54f07c1df4" />

 Are you fine with the second option or do you still think the first is better? (@PepperLola)

AlexD717 committed:
Configure Assets Multiple Tabs

AlexD717 committed:
Register Commands Formatting

AlexD717 committed:
Removed Useless Command & Black Border

azaleacolburn commented:
I like the last option best personally

AlexD717 committed:
Merge branch 'dev' into alexey/2079/command-tool

AlexD717 committed:
Disable Autocomplete

Autocomplete could sometimes get in the way, and be annoying.

PepperLola reviewed (changes requested):
I like the separator but I think it would make sense either to have the border around the whole thing or just have a separator between the input and the options (only the top of the input), and maybe then get rid of the border radius on the top left and right.

PepperLola committed:
Merge branch 'dev' into alexey/2079/command-tool

AlexD717 committed:
No radius in text field top edges

PepperLola reviewed (changes requested):
When I added a bunch of new commands, they weren't all accessible since it was only including enough to fit. I think the list should have a max height but should be scrollable and include all the registered commands
<img width="1200" alt="Image" src="https://github.com/user-attachments/assets/5f13df2c-295f-496c-b892-b207663d9aa1" />

AlexD717 commented:
> I like the separator but I think it would make sense either to have the border around the whole thing

The border is around the whole thing right now

> When I added a bunch of new commands, they weren't all accessible since it was only including enough to fit. I think the list should have a max height but should be scrollable and include all the registered commands <img alt="Image" width="1200" src="https://private-user-images.githubusercontent.com/20529380/481118935-5f13df2c-295f-496c-b892-b207663d9aa1.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU4OTI0NDksIm5iZiI6MTc1NTg5MjE0OSwicGF0aCI6Ii8yMDUyOTM4MC80ODExMTg5MzUtNWYxM2RmMmMtMjk1Zi00OTZjLWI4OTItYjIwNzY2M2Q5YWExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODIyVDE5NDkwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU2NzdhN2QzNTQ1NjNjOWZkNzY4Zjc1ZDA3NjMwOWZhMTVlYjBmNGFlNmZmYjU3ZjYyYTQ2ZWE4ZWNjODJmYjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.OcbCekdz3dwwHAQwAyu0Q4_f8Lr5PnaXykPR3BbHx54">

If you start typing it very quickly filters them out

PepperLola commented:
> If you start typing it very quickly filters them out

That's true, but I think you should be able to scroll through all of them to see what options are available. It's true that you can filter some out so I guess it's not super important to include in this PR but I definitely think it would be worthwhile to add.

EDIT: on second thought, I think it should be included in this PR as it's pretty core functionality and I think necessary for a good user experience

AlexD717 committed:
Scrollable Command Palette

PepperLola reviewed (approved):
I think it looks good. It would be nice if the scrolling were smooth, and there were a couple bugs I couldn't figure out how to consistently reproduce (sometimes it would always scroll the selected option to the second slot from the top, regardless of whether I was scrolling up or down). I think it's in a good state to be merged, keeping these fixes in mind for the future.

PepperLola committed:
Merge branch 'dev' into alexey/2079/command-tool

PepperLola committed:
Merge branch 'dev' into alexey/2079/command-tool
